components/mediation-monitor/mediation-data-publisher/org.wso2.carbon.prometheus.publisher/src/main/java/org/wso2/carbon/prometheus/publisher/publisher/MetricCollector.java

```diff
@@ -0,0 +1,358 @@
+                for (PrometheusMetric metric : metricList) {
+                    metrics.add(PrometheusMetric.getFormattedMetric(metric));
+                }
+
+                return metrics;
+            }
+
+        } catch (MalformedObjectNameException e) {
+            LOGGER.error("Error in retrieving metric data from JMX MBeans: " + e);
+        }
+        return Collections.emptyList();
+    }
+
+    private List<PrometheusMetric> scrapeBean(MBeanServer beanServer, ObjectName beanName, Map<ObjectName, LinkedHashMap<String, String>> keyPropertiesPerBean) {
+
+        try {
+            // get attributes of each bean
+            MBeanInfo beanInfo = beanServer.getMBeanInfo(beanName);
+            MBeanAttributeInfo[] beanAttributes = beanInfo.getAttributes();
+            Map<String, MBeanAttributeInfo> attributeMap = new HashMap<>();
+            for (MBeanAttributeInfo attribute : beanAttributes) {
+                if (attribute.isReadable()) {
+                    attributeMap.put(attribute.getName(), attribute);
+                }
+            }
+            AttributeList attributes = beanServer.getAttributes(beanName,
+                    attributeMap.keySet().toArray(new String[0]));
+
+            List<PrometheusMetric> metrics = new ArrayList<>();
+
+            // process bean attributes
+            for (Attribute attribute : attributes.asList()) {
+                MBeanAttributeInfo attributeInfo = attributeMap.get(attribute.getName());
+                PrometheusMetric metric = processBeanValue(toSnakeAndLowerCase(beanName.getDomain()),
+                        toSnakeAndLowerCase(attributeInfo.getName()),
+                        attributeInfo.getDescription(), attribute.getValue(), getKeyPropertyList(keyPropertiesPerBean, beanName));
+                if (metric != null) {
+                    metrics.add(metric);
+                }
+            }
+            return metrics;
+        } catch (ReflectionException | IntrospectionException | InstanceNotFoundException e) {
+            LOGGER.error("Error in scraping bean - " + beanName.toString());
+        }
+        return Collections.emptyList();
+    }
+
+    /**
+     * Convert metric name into Prometheus format
+     *
+     * @param attrName Metric name
+     * @return Formatted name
+     */
+    private String toSnakeAndLowerCase(String attrName) {
+
+        if (attrName == null || attrName.isEmpty()) {
+            return attrName;
+        }
+        char firstChar = attrName.subSequence(0, 1).charAt(0);
+        boolean prevCharIsUpperCaseOrUnderscore = Character.isUpperCase(firstChar) || firstChar == '_';
+        StringBuilder resultBuilder =
+                new StringBuilder(attrName.length()).append(Character.toLowerCase(firstChar));
+        for (char attrChar : attrName.substring(1).toCharArray()) {
+            boolean charIsUpperCase = Character.isUpperCase(attrChar);
+            if (attrChar == '.') {
```

Comment:
we can merge these 2 conditions since the operation is the same in both cases ( resultBuilder.append("_"); )